README.md
# <img src="https://github.com/pip-services/pip-services/raw/master/design/Logo.png" alt="Pip.Services Logo" style="max-width:30%"> <br/> Redis components for Node.js

This component library is a part of [Pip.Services](https://github.com/pip-services/pip-services) project.
It contains Redis components: RedisCache and RedisLock

Quick Links:

* [Downloads](https://github.com/pip-services-node/pip-services-redis-node/blob/master/doc/Downloads.md)
* [API Reference](https://rawgit.com/pip-services-node/pip-services-redis-node/master/doc/api/index.html)
* [Building and Testing](https://github.com/pip-services-node/pip-services-redis-node/blob/master/doc/Development.md)
* [Contributing](https://github.com/pip-services-node/pip-services-redis-node/blob/master/doc/Development.md/#contrib)

## Acknowledgements

The library is created and maintained by **Sergey Seroukhov**.


CHANGELOG.md
# Redis components for Pip.Services in Node.js Changelog

## <a name="3.0.0"></a> 3.0.0 (2018-08-21) 

### Breaking Changes
* Moved to a separate package

## <a name="1.2.0"></a> 1.2.0 (2018-08-10) 

### Features
* Added RedisCache
* Added RedisLock

## <a name="1.1.0"></a> 1.1.0 (2018-03-26) 

### Features
* Added PrometheusCounters and PrometheusMetricsService
* Added labels to PrometheusCounters and PrometheusMetricsService

## <a name="1.0.0"></a> 1.0.0 (2018-03-20) 

### Features
* **memcached** Added MemcachedCache
* **memcached** Added MemcachedLock
* **fluentd** Added FluentdLogger
* **elasticsearch** Added ElasticSearchLogger



test/lock/RedisLock.test.ts
import { ConfigParams } from 'pip-services-commons-node';

import { RedisLock } from '../../src/lock/RedisLock';
import { LockFixture } from '../fixtures/LockFixture';

suite('RedisLock', ()=> {
    var _lock: RedisLock;
    var _fixture: LockFixture;

    setup((done) => {
        let host = process.env['REDIS_SERVICE_HOST'] || 'localhost';
        let port = process.env['REDIS_SERVICE_PORT'] || 6379;

        _lock = new RedisLock();

        let config = ConfigParams.fromTuples(
            'connection.host', host,
            'connection.port', port
        );
        _lock.configure(config);

        _fixture = new LockFixture(_lock);

        _lock.open(null, done);
    });

    teardown((done) => {
        _lock.close(null, done);
    });

    test('Try Acquire Lock', (done) => {
        _fixture.testTryAcquireLock(done);
    });

    test('Acquire Lock', (done) => {
        _fixture.testAcquireLock(done);
    });

    test('Release Lock', (done) => {
        _fixture.testReleaseLock(done);
    });

});


test/fixtures/LockFixture.ts
let async = require('async');
let assert = require('chai').assert;

import { ILock } from 'pip-services-components-node';

let LOCK1: string = "lock_1";
let LOCK2: string = "lock_2";
let LOCK3: string = "lock_3";

export class LockFixture {
    private _lock: ILock;

    public constructor(lock: ILock) {
        this._lock = lock;
    }

    public testTryAcquireLock(done: (err: any) => void): void {
        async.series([
            // Try to acquire lock for the first time
            (callback) => {
                this._lock.tryAcquireLock(null, LOCK1, 3000, (err, result) => {
                    assert.isNull(err || null);
                    assert.isTrue(result);
                    callback();
                });
            },
            // Try to acquire lock for the second time
            (callback) => {
                this._lock.tryAcquireLock(null, LOCK1, 3000, (err, result) => {
                    assert.isNull(err || null);
                    assert.isFalse(result);
                    callback();
                });
            },
            // Release the lock
            (callback) => {
                this._lock.releaseLock(null, LOCK1, callback);
            },
            // Try to acquire lock for the third time
            (callback) => {
                this._lock.tryAcquireLock(null, LOCK1, 3000, (err, result) => {
                    assert.isNull(err || null);
                    assert.isTrue(result);
                    callback();
                });
            }
        ], (err) => {
            this._lock.releaseLock(null, LOCK1);
            done(err);
        });
    }

    public testAcquireLock(done: (err: any) => void): void {
        async.series([
            // Acquire lock for the first time
            (callback) => {
                this._lock.acquireLock(null, LOCK2, 3000, 1000, (err) => {
                    assert.isNull(err || null);
                    callback();
                });
            },
            // Acquire lock for the second time
            (callback) => {
                this._lock.acquireLock(null, LOCK2, 3000, 1000, (err) => {
                    assert.isNotNull(err || null);
                    callback();
                });
            },
            // Release the lock
            (callback) => {
                this._lock.releaseLock(null, LOCK2, callback)
            },
            // Acquire lock for the third time
            (callback) => {
                this._lock.acquireLock(null, LOCK2, 3000, 1000, (err) => {
                    assert.isNull(err || null);
                    callback();
                });
            },
        ], (err) => {
            this._lock.releaseLock(null, LOCK2);
            done(err);
        });
    }

    public testReleaseLock(done: (err: any) => void): void {
        async.series([
            // Acquire lock for the first time
            (callback) => {
                this._lock.tryAcquireLock(null, LOCK3, 3000, (err, result) => {
                    assert.isNull(err || null);
                    assert.isTrue(result);
                    callback();
                });
            },
            // Release the lock for the first time
            (callback) => {
                this._lock.releaseLock(null, LOCK3, callback)
            },
            // Release the lock for the second time
            (callback) => {
                this._lock.releaseLock(null, LOCK3, callback)
            }
        ], done);
    }
    
}

test/cache/RedisCache.test.ts
let process = require('process');
let assert = require('chai').assert;
let async = require('async');

import { Descriptor } from 'pip-services-commons-node';
import { ConfigParams } from 'pip-services-commons-node';

import { RedisCache } from '../../src/cache/RedisCache';
import { CacheFixture } from '../fixtures/CacheFixture';

suite('RedisCache', ()=> {
    let _cache: RedisCache;
    let _fixture: CacheFixture;

    setup((done) => {
        let host = process.env['REDIS_SERVICE_HOST'] || 'localhost';
        let port = process.env['REDIS_SERVICE_PORT'] || 6379;

        _cache = new RedisCache();

        let config = ConfigParams.fromTuples(
            'connection.host', host,
            'connection.port', port
        );
        _cache.configure(config);

        _fixture = new CacheFixture(_cache);

        _cache.open(null, done);
    });

    teardown((done) => {
        _cache.close(null, done);
    });

    test('Store and Retrieve', (done) => {
        _fixture.testStoreAndRetrieve(done);
    });    

    test('Retrieve Expired', (done) => {
        _fixture.testRetrieveExpired(done);
    });    

    test('Remove', (done) => {
        _fixture.testRemove(done);
    });    
    
});


src/cache/RedisCache.ts
const _ = require('lodash');
const async = require('async');

import { ConfigParams } from 'pip-services-commons-node';
import { IConfigurable } from 'pip-services-commons-node';
import { IReferences } from 'pip-services-commons-node';
import { IReferenceable } from 'pip-services-commons-node';
import { IOpenable } from 'pip-services-commons-node';
import { InvalidStateException } from 'pip-services-commons-node';
import { ConfigException } from 'pip-services-commons-node';
import { ConnectionParams } from 'pip-services-components-node';
import { ConnectionResolver } from 'pip-services-components-node';
import { CredentialParams } from 'pip-services-components-node';
import { CredentialResolver } from 'pip-services-components-node';
import { ICache } from 'pip-services-components-node';

export class RedisCache implements ICache, IConfigurable, IReferenceable, IOpenable {
    private _connectionResolver: ConnectionResolver = new ConnectionResolver();
    private _credentialResolver: CredentialResolver = new CredentialResolver();
    
    private _timeout: number = 30000;
    private _retries: number = 3;

    private _client: any = null;

    public constructor() {}

    public configure(config: ConfigParams): void {
        this._connectionResolver.configure(config);
        this._credentialResolver.configure(config);

        this._timeout = config.getAsIntegerWithDefault('options.timeout', this._timeout);
        this._retries = config.getAsIntegerWithDefault('options.retries', this._retries);
    }

    public setReferences(references: IReferences): void {
        this._connectionResolver.setReferences(references);
        this._credentialResolver.setReferences(references);
    }

    public isOpened(): boolean {
        return this._client;
    }

    public open(correlationId: string, callback: (err: any) => void): void {
        let connection: ConnectionParams;
        let credential: CredentialParams;

        async.series([
            (callback) => {
                this._connectionResolver.resolve(correlationId, (err, result) => {
                    connection = result;
                    if (err == null && connection == null)
                        err = new ConfigException(correlationId, 'NO_CONNECTION', 'Connection is not configured');
                    callback(err);
                });
            },
            (callback) => {
                this._credentialResolver.lookup(correlationId, (err, result) => {
                    credential = result;
                    callback(err);
                });
            },
            (callback) => {
                let options: any = {
                    // connect_timeout: this._timeout,
                    // max_attempts: this._retries,
                    retry_strategy: (options) => { return this.retryStrategy(options); }
                };
                
                if (connection.getUri() != null) {
                    options.url = connection.getUri();
                } else {                    
                    options.host = connection.getHost() || 'localhost';
                    options.port = connection.getPort() || 6379;
                }

                if (credential != null) {
                    options.password = credential.getPassword();
                }
    
                let redis = require('redis');
                this._client = redis.createClient(options);
    
                if (callback) callback(null);    
            }
        ], callback);
    }

    public close(correlationId: string, callback: (err: any) => void): void {
        if (this._client != null) {
            this._client.quit(((err) => {
                this._client = null;    
                if (callback) callback(err);
            }));
        } else {
            if (callback) callback(null);
        }
    }

    private checkOpened(correlationId: string, callback: any): boolean {
        if (!this.isOpened()) {
            let err = new InvalidStateException(correlationId, 'NOT_OPENED', 'Connection is not opened');
            callback(err, null);
            return false;
        }
        
        return true;
    }
    
    private retryStrategy(options: any): any {
        if (options.error && options.error.code === 'ECONNREFUSED') {
            // End reconnecting on a specific error and flush all commands with
            // a individual error
            return new Error('The server refused the connection');
        }
        if (options.total_retry_time > this._timeout) {
            // End reconnecting after a specific timeout and flush all commands
            // with a individual error
            return new Error('Retry time exhausted');
        }
        if (options.attempt > this._retries) {
            // End reconnecting with built in error
            return undefined;
        }
        // reconnect after
        return Math.min(options.attempt * 100, 3000);
    }

    public retrieve(correlationId: string, key: string,
        callback: (err: any, value: any) => void): void {
        if (!this.checkOpened(correlationId, callback)) return;

        this._client.get(key, callback);
    }

    public store(correlationId: string, key: string, value: any, timeout: number,
        callback: (err: any) => void): void {
        if (!this.checkOpened(correlationId, callback)) return;

        this._client.set(key, value, 'PX', timeout, callback);
    }

    public remove(correlationId: string, key: string,
        callback: (err: any) => void) {
        if (!this.checkOpened(correlationId, callback)) return;

        this._client.del(key, callback);
    }
    
}

src/lock/RedisLock.ts
const _ = require('lodash');
const async = require('async');

import { ConfigParams } from 'pip-services-commons-node';
import { IConfigurable } from 'pip-services-commons-node';
import { IReferences } from 'pip-services-commons-node';
import { IReferenceable } from 'pip-services-commons-node';
import { IOpenable } from 'pip-services-commons-node';
import { IdGenerator } from 'pip-services-commons-node';
import { InvalidStateException } from 'pip-services-commons-node';
import { ConfigException } from 'pip-services-commons-node';
import { ConnectionParams } from 'pip-services-components-node';
import { ConnectionResolver } from 'pip-services-components-node';
import { CredentialParams } from 'pip-services-components-node';
import { CredentialResolver } from 'pip-services-components-node';
import { Lock } from 'pip-services-components-node';

export class RedisLock extends Lock implements IConfigurable, IReferenceable, IOpenable {
    private _connectionResolver: ConnectionResolver = new ConnectionResolver();
    private _credentialResolver: CredentialResolver = new CredentialResolver();
    
    private _lock: string = IdGenerator.nextLong();
    private _timeout: number = 30000;
    private _retries: number = 3;

    private _client: any = null;

    public configure(config: ConfigParams): void {
        this._connectionResolver.configure(config);
        this._credentialResolver.configure(config);

        this._timeout = config.getAsIntegerWithDefault('options.timeout', this._timeout);
        this._retries = config.getAsIntegerWithDefault('options.retries', this._retries);
    }

    public setReferences(references: IReferences): void {
        this._connectionResolver.setReferences(references);
        this._credentialResolver.setReferences(references);
    }

    public isOpened(): boolean {
        return this._client;
    }

    public open(correlationId: string, callback: (err: any) => void): void {
        let connection: ConnectionParams;
        let credential: CredentialParams;

        async.series([
            (callback) => {
                this._connectionResolver.resolve(correlationId, (err, result) => {
                    connection = result;
                    if (err == null && connection == null)
                        err = new ConfigException(correlationId, 'NO_CONNECTION', 'Connection is not configured');
                    callback(err);
                });
            },
            (callback) => {
                this._credentialResolver.lookup(correlationId, (err, result) => {
                    credential = result;
                    callback(err);
                });
            },
            (callback) => {
                let options: any = {
                    // connect_timeout: this._timeout,
                    // max_attempts: this._retries,
                    retry_strategy: (options) => { return this.retryStrategy(options); }
                };
                
                if (connection.getUri() != null) {
                    options.url = connection.getUri();
                } else {                    
                    options.host = connection.getHost() || 'localhost';
                    options.port = connection.getPort() || 6379;
                }

                if (credential != null) {
                    options.password = credential.getPassword();
                }
    
                let redis = require('redis');
                this._client = redis.createClient(options);
    
                if (callback) callback(null);    
            }
        ], callback);
    }

    public close(correlationId: string, callback: (err: any) => void): void {
        if (this._client != null) {
            this._client.quit(((err) => {
                this._client = null;    
                if (callback) callback(err);
            }));
        } else {
            if (callback) callback(null);
        }
    }

    private checkOpened(correlationId: string, callback: any): boolean {
        if (!this.isOpened()) {
            let err = new InvalidStateException(correlationId, 'NOT_OPENED', 'Connection is not opened');
            callback(err, null);
            return false;
        }
        
        return true;
    }
    
    private retryStrategy(options: any): any {
        if (options.error && options.error.code === 'ECONNREFUSED') {
            // End reconnecting on a specific error and flush all commands with
            // a individual error
            return new Error('The server refused the connection');
        }
        if (options.total_retry_time > this._timeout) {
            // End reconnecting after a specific timeout and flush all commands
            // with a individual error
            return new Error('Retry time exhausted');
        }
        if (options.attempt > this._retries) {
            // End reconnecting with built in error
            return undefined;
        }
        // reconnect after
        return Math.min(options.attempt * 100, 3000);
    }

    public tryAcquireLock(correlationId: string, key: string, ttl: number,
        callback: (err: any, result: boolean) => void): void {
        if (!this.checkOpened(correlationId, callback)) return;

        this._client.set(key, this._lock, 'NX', 'PX', ttl, (err, result) => {
            callback(err, result == "OK");
        });
    }

    public releaseLock(correlationId: string, key: string,
        callback?: (err: any) => void): void {
        if (!this.checkOpened(correlationId, callback)) return;

        // Start transaction on key
        this._client.watch(key, (err) => {
            if (err) {
                if (callback) callback(err);
                return;
            }

            // Read and check if lock is the same
            this._client.get(key, (err, result) => {
                if (err) {
                    if (callback) callback(err);
                    return;                    
                }

                // Remove the lock if it matches
                if (result == this._lock) {
                    this._client.multi()
                        .del(key)
                        .exec(callback);
                }
                // Cancel transaction if it doesn't match
                else {
                    this._client.unwatch(callback);
                }
            })
        });
    }    
}